python/mlir_structured/dialects/indexing.py

```diff
@@ -29,90 +46,199 @@
 # The classes also provide some convenience factory methods and properties for accessing metadata (data type, shape, etc.).
 
 
-class Scalar(ScalarValue):
-
+def infer_mlir_type(py_val: Union[int, float, bool, np.ndarray]) -> Type:
+  if isinstance(py_val, bool):
+    return IntegerType.get_signless(1)
+  elif isinstance(py_val, int):
+    return IntegerType.get_signless(64)
+  elif isinstance(py_val, float):
+    return F64Type.get()
+  elif isinstance(py_val, np.ndarray):
+    return {
+        np.int8: IntegerType.get_signless(8),
+        np.int16: IntegerType.get_signless(16),
+        np.int32: IntegerType.get_signless(32),
+        np.int64: IntegerType.get_signless(64),
+        np.uintp: IndexType.get(),
+        np.float16: F16Type.get(),
+        np.float32: F32Type.get(),
+        np.float64: F64Type.get(),
+    }[py_val.dtype.type]
+  else:
+    raise NotImplementedError(
+        f"Unsupported Python value {py_val=} with type {type(py_val)}")
+
+
+def constant(
+    cst: Union[int, float, bool, np.ndarray],
+    type: Optional[Type] = None,
+    index: Optional[bool] = None,
+) -> arith.ConstantOp:
+  assert isinstance(cst, (int, float, bool, np.ndarray))
+  if index is not None and index:
+    type = IndexType.get()
+  elif type is None:
+    type = infer_mlir_type(cst)
+  assert type is not None
+
+  if isinstance(cst, np.ndarray):
+    type = RankedTensorType.get(cst.shape, type)
+    cst = DenseElementsAttr.get(
+        cst,
+        type=type,
+    )
+
+  cst = arith.ConstantOp(type, cst)
+
+  return cst
+
+
+# TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
+class ArithValueMeta(type(Value)):
+
+  def __call__(cls, *args, **kwargs):
+    if len(args) != 1:
+      raise ValueError("Only one non-kw arg supported.")
+    arg = args[0]
+    if isinstance(arg, (OpView, Operation, Value)):
+      if isinstance(arg, (Operation, OpView)):
+        assert len(arg.results) == 1
+      val = get_op_result_or_value(arg)
+      arg_copy = None
+    elif isinstance(arg, (int, float, bool, np.ndarray)):
+      dtype = kwargs.get("dtype")
+      index = kwargs.get("index")
+      arg_copy = deepcopy(arg)
```

Comment:
In particular this is for `np.ndarray`, i.e., decouple from the numpy value since `DenseElementsAttr.get(arr)` takes a [view](https://github.com/llvm/llvm-project/blob/040a41a852933d3d1b855aebc8b054baa60f61e2/mlir/lib/Bindings/Python/IRAttributes.cpp#L556) rather than making a copy.